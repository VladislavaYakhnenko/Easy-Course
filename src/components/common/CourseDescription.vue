<template>
  <p class="font-montserrat font-normal text-gray-700 dark:text-indigo-100 max-lg:text-sm max-md:text-xs">
    {{ description }}
  </p>
</template>

<script setup lang="ts">
interface Props {
  description: string
}

const props = defineProps<Props>()
</script>

<style scoped>

</style>