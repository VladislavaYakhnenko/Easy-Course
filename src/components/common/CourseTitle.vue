<template>
  <h2 class="font-montserrat font-semibold text-gray-800 dark:text-indigo-50 text-3xl max-xl:text-2xl max-lg:text-xl max-md:text-lg">
    {{ title }}
  </h2>
</template>

<script setup lang="ts">
interface Props {
  title: string
}

const props = defineProps<Props>()
</script>

<style scoped>

</style>