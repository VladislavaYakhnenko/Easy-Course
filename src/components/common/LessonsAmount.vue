<template>
  <span class="font-montserrat font-semibold text-gray-500 dark:text-gray-300 max-xl:text-sm">
      {{ lessonsCount }} lessons
  </span>
</template>

<script setup lang="ts">
interface Props {
  lessonsCount: number
}

const props = defineProps<Props>()
</script>

<style scoped>

</style>