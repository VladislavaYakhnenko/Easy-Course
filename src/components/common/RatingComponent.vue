<template>
  <div class="flex items-center gap-1.5">
    <span class="font-montserrat font-semibold text-gray-500 text-xl max-lg:text-lg max-md:text-base">
      {{ getRating }}
    </span>
    <StarIcon/>
  </div>
</template>

<script setup>
import StarIcon from "@/components/icons/StarIcon.vue";
import {computed} from "vue";

const props = defineProps({
  rating: {
    type: Number,
    required: true,
  }
})

const getRating = computed(() => {
  return Number.isInteger(props.rating) ? props.rating + '.0' : props.rating;
})
</script>

<style scoped>

</style>