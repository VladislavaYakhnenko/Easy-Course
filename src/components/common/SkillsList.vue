<template>
  <div class="w-full flex flex-col gap-3 font-montserrat">
    <span class="font-semibold text-gray-700 dark:text-indigo-100 text-xl max-xl:text-base max-md:text-sm">
      Skills:
    </span>
    <div v-if="skills?.length">
      <div
          class="flex items-center gap-2"
          v-for="skill in skills"
          :key="skill"
      >
        <div class="w-2 h-2 rounded-full bg-indigo-600  dark:bg-indigo-400 max-lg:w-1 max-lg:h-1"/>
        <li class="font-normal list-none text-gray-700 dark:text-indigo-100 max-lg:text-sm max-md:text-xs">
          {{ skill }}
        </li>
      </div>
    </div>
    <span
        v-else
        class="font-normal text-gray-700 dark:text-indigo-100"
    >
      No additional skills
    </span>
  </div>
</template>

<script setup lang="ts">
interface Props {
  skills?: string[]
}

const props = defineProps<Props>();
</script>

<style scoped>

</style>