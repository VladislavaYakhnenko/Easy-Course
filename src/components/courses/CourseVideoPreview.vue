<template>
  <div
      class="rounded-2xl shadow-md ring-white ring-2 overflow-hidden w-full aspect-video"
      @mouseover="playVideo"
      @mouseleave="pauseVideo"
  >
    <video
        class="w-full h-full object-cover"
        ref="videoRef"
        loop
        :poster="previewImageLink + '/cover.webp'"
        preload="none"
        muted
    />
  </div>
</template>

<script setup>
import {onMounted, ref} from "vue";
import Hls from "hls.js/dist/hls.min";

const videoRef = ref(null);

const props = defineProps({
  previewImageLink: {
    type: String,
    required: true,
  },
  videoLink: {
    type: String,
    required: true,
  }
})

const playVideo = () => {
  if (videoRef.value) {
    videoRef.value.play();
  }
};

const pauseVideo = () => {
  if (videoRef.value) {
    videoRef.value.pause();
  }
};

onMounted(() => {
  if (Hls.isSupported()) {
    const hls = new Hls();
    hls.loadSource(props.videoLink);
    hls.attachMedia(videoRef.value);
  } else if (videoRef.value.canPlayType('application/vnd.apple.mpegurl')) {
    videoRef.value.src = props.videoLink;
  }
});

</script>

<style scoped>

</style>