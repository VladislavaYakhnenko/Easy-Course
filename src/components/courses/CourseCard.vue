<template>
  <RouterLink :to="{ name: 'about-course', params: {courseId: course?.id }}">
    <article class="w-full h-full bg-indigo-50 flex flex-col gap-4 justify-between p-6 rounded-2xl border-2 border-indigo-100 cursor-pointer duration-200 hover:scale-[101%] hover:shadow-2xl">
      <div class="w-full flex flex-col gap-4">
        <CourseTitle :title="course?.title"/>
        <CourseVideoPreview :previewImageLink="course?.previewImageLink"
                            :videoLink="course?.meta?.courseVideoPreview?.link"
        />
        <CourseDescription :description="course?.description"/>
        <SkillsList :skills="course?.meta?.skills"/>
      </div>
      <div class="w-full flex items-center justify-between">
        <LessonsAmount :lessonsCount="course?.lessonsCount"/>
        <RatingComponent :rating="course?.rating"/>
      </div>
    </article>
  </RouterLink>
</template>

<script setup>
import CourseVideoPreview from "@/components/courses/CourseVideoPreview.vue";
import CourseTitle from "@/components/common/CourseTitle.vue";
import CourseDescription from "@/components/common/CourseDescription.vue";
import SkillsList from "@/components/common/SkillsList.vue";
import LessonsAmount from "@/components/common/LessonsAmount.vue";
import RatingComponent from "@/components/common/RatingComponent.vue";

const props = defineProps({
  course: {
    type: Object,
    required: true,
  }
})

</script>

<style scoped>

</style>