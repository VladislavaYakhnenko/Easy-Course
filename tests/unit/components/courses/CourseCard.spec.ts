import { shallowMount } from '@vue/test-utils';
import { expect, describe, it } from 'vitest';
import { RouterLink } from "vue-router";
import CourseCard from "@/components/courses/CourseCard.vue";
import CourseTitle from "@/components/common/CourseTitle.vue";
import CourseVideoPreview from "@/components/courses/CourseVideoPreview.vue";
import CourseDescription from "@/components/common/CourseDescription.vue";
import SkillsList from "@/components/common/SkillsList.vue";
import LessonsAmount from "@/components/common/LessonsAmount.vue";
import RatingComponent from "@/components/common/RatingComponent.vue";

const mockCourse = {
  id: "1",
  title: "Test Course",
  previewImageLink: "https://photo.com/image.png",
  description: "Test Course Description",
  lessonsCount: 5,
  rating: 4,
  meta: {
    skills: ["JavaScript", "Vue.js"],
    courseVideoPreview: {
      link: "https://video.com/video.mp4"
    }
  }
};

describe("CoursePreview", () => {
  it("renders the course preview", () => {
    const wrapper = shallowMount(CourseCard, {
      props: { course: mockCourse },
      global: {
        components: {
          RouterLink,
          CourseTitle,
          CourseVideoPreview,
          CourseDescription,
          SkillsList,
          LessonsAmount,
          RatingComponent
        }
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});