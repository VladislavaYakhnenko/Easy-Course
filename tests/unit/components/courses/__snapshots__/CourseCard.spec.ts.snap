// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoursePreview > renders the course preview 1`] = `"<router-link-stub to=\\"[object Object]\\" replace=\\"false\\" custom=\\"false\\" ariacurrentvalue=\\"page\\"></router-link-stub>"`;
